docs/library/ubluetooth.rst

```diff
@@ -493,6 +514,64 @@ device name from the device information service).
     peripheral initiating the MTU exchange. NimBLE works for both roles.
 
 
+L2CAP connection-oriented-channels (CoC)
+----------------------------------------
+
+    This feature allows for socket-like data exchange between two BLE devices.
+    Once the devices are connected via GAP, either device can listen for the
+    other to connect.
+
+    **Note:** This is currently only supported when using the NimBLE stack on
+    STM32 and Unix (not ESP32). Only one L2CAP CoC may be active at a given time
+    (i.e. you cannot connect while listening).
+
+.. method:: BLE.l2cap_coc_listen(psm, mtu, /)
+
+    Start listening for incoming L2CAP CoC requests on the specified *psm*.
+
+    The *mtu* specified will be the maximum payload size that this device can
+    transmit (which may be different to the MTU used in the other direction).
+
+    When a remote device initiates a connection, the ``_IRQ_L2CAP_COC_ACCEPT``
+    event will be raised, followed by the ``_IRQ_L2CAP_COC_CONNECT`` event. Both
+    events include a "channel handle", and the connect event includes the local
+    and remote MTU.
+
+.. method:: BLE.l2cap_coc_connect(conn_handle, psm, mtu, /)
+
+    Connect to a listening peer on the specified *psm*.
+
+    The *mtu* specified will be the maximum payload size that this device can
+    transmit (which may be different to the MTU used in the other direction).
+
+    On successful connection, the the ``_IRQ_L2CAP_COC_CONNECT`` event will be
+    raised, including a "channel handle" and the local and remote MTU.
+
+.. method:: BLE.l2cap_coc_disconnect(chan_handle, /)
+
+    Disconnect an active L2CAP CoC with the specified *chan_handle*.
+
+.. method:: BLE.l2cap_coc_send(chan_handle, buf, /)
+
+    Send the specified *buf* (which must support the buffer protocol) on the
+    L2CAP CoC identified by *chan_handle*.
+
+    This will return ``True`` if the channel is now "stalled", which means that
+    :meth:`l2cap_coc_send <BLE.l2cap_coc_send>` must not be called again until
+    the ``_IRQ_L2CAP_COC_TX_UNSTALLED`` event is received.
+
+    The buffer must always be smaller than the MTU.
```

Comment:
is the entire buffer always sent (or at least queued for sending)?  does the user need to retain and not modify the buffer if the send stalled?

Reply:
Yes, the idea is that the buffer is immediately copied (in our code) into a NimBLE mbuf, and then NimBLE will send the whole thing, only failing if something goes wrong with mbuf management. To the best of my understanding, I have given NimBLE's mbuf's enough room such that it should be impossible for this to happen (i.e. there's always a full MTU worth of mbuf space for tx, and two MTU's worth for RX, and these are shared in the same pool, so for a send-only client, there's three MTU's worth of room).

However, this is why the `status` field exists on the SEND_READY event. If it's non-zero, then at least some data was lost. (But... this shouldn't happen.)